### PR TITLE
Update README.md

### PR DESCRIPTION
**Pull request by cooperlab**
Added details for running executables.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -28,3 +28,127 @@ After matching the slide filenames against the sample IDs from the clinical data
...
 
 *Note: region extraction depends on the [OpenSlide](http://openslide.org/) library. This library can be installed locally, and is also available in the Docker image described below.*
+
+# Executables for training and testing models
+
+The Docker container provides executables for training SCNN/GSCNN models and for evaluating the accuracy of these models. Both of these executables consume:
+* ROI .png files
+* A .csv file describing the patient IDs, event times, censoring (0 for uncensored, 1 for right-censored), patient indexes (integer indexes assigned to each patient), and any genomic variables. **_Note that the columns for genomic data should be organized in a way that the first coloumns are allocated for those which will NOT get normalized and the latter ones allocated for those that have to get normnalized._** The default normalization method in our model is Z score normalization. An example file is provided in this repository.
```
Comment (cooperlab) [inline README.md]: Spelling

Comment (cooperlab) [inline README.md]: Can we just have people responsible for their own normalization?

File: README.md [hunk 2]
```diff
@@ -28,3 +28,127 @@ After matching the slide filenames against the sample IDs from the clinical data
...
+                      [--ngf NGF] [--gn GN] [--ns NS] [--nm NM]
+
+Arguments for Training the SCNN/GSCNN model
+
+optional arguments:
+  -h, --help  show this help message and exit
+  -m M        image_only or image_genome; image_only for SCNN and image_genome
+              for GSCNN. The default is image_only.
+  -p P        Path to the Train/Test patient list. The default path is ./inputs/ImageSplits.txt
+  -f F        Path to the features data. The default path is ./inputs/all_dataset.csv
```
Comment (cooperlab) [inline README.md]: This is clinical outcomes and genomic features? Need specifics.

File: README.md [hunk 3]
```diff
@@ -28,3 +28,127 @@ After matching the slide filenames against the sample IDs from the clinical data
...
+                      [--lr LR] [--me ME] [--kp KP] [--bs BS] [--ic IC]
+                      [--ngf NGF] [--gn GN] [--ns NS] [--nm NM]
+
+Arguments for Training the SCNN/GSCNN model
+
+optional arguments:
+  -h, --help  show this help message and exit
+  -m M        image_only or image_genome; image_only for SCNN and image_genome
+              for GSCNN. The default is image_only.
+  -p P        Path to the Train/Test patient list. The default path is ./inputs/ImageSplits.txt
```
Comment (cooperlab) [inline README.md]: Do we use this anymore? I don't think so. The -i argument seems to take care of this.

File: README.md [hunk 4]
```diff
@@ -28,3 +28,127 @@ After matching the slide filenames against the sample IDs from the clinical data
...
+  -f F        Path to the features data. The default path is ./inputs/all_dataset.csv
+  -i I        Path to the folder containing ROI .pngs. The default path is ../images/train
+  -r R        Path where the training outputs should be generated. The default path is ./train_results
```
Comment (cooperlab) [inline README.md]: 'training outputs' need specifics

File: README.md [hunk 5]
```diff
@@ -28,3 +28,127 @@ After matching the slide filenames against the sample IDs from the clinical data
...
+  -t T        Path where training meta files should be generated. The default path is ./checkpoints
```
Comment (cooperlab) [inline README.md]: 'meta files' may not be clear to people

File: README.md [hunk 6]
```diff
@@ -28,3 +28,127 @@ After matching the slide filenames against the sample IDs from the clinical data
...
+  -d D        Path where binary files for training should be generated (these can be quite large and
+              should be cleaned). The default path is ./tmp
+  --lr LR     Initial learning rate. Default value = 0.001.
+  --me ME     Max number of epochs. Default value = 100.
+  --kp KP     Keeping probability for Test time dropout. Default value = 0.95.
+  --bs BS     batch size. Default value = 14.
+  --ic IC     indexes column in the number in the ./inputs/all_dataset.csv
```
Comment (cooperlab) [inline README.md]: I think we need a note on the role of indexes somewhere. People may wonder why you need to assign an integer to each patient when there is a natural order within the outcomes/genomics .csv file.

File: README.md [hunk 7]
```diff
@@ -28,3 +28,127 @@ After matching the slide filenames against the sample IDs from the clinical data
...
+              file, starting from 0. Default value = 1.
+  --ngf NGF   Number of genomic features. The default is 2, considering the
+              case we have just "codeletion" and "idh mutation" as genomic
+              features.
+  --gn GN     Z Normalization of the geomic values; y for yes and n for no.
```
Comment (cooperlab) [inline README.md]: I think it's ok to just mention we use z-score in the paper, and leave it to people to normalize their own features. It's much easier, and they may want to use a different method.

File: README.md [hunk 8]
```diff
@@ -28,3 +28,127 @@ After matching the slide filenames against the sample IDs from the clinical data
...
+              the -f argument file, starting from 0 index. Default value = 6.
+              Note: The columns we want to normalize should get assigned
+              as the last columns of the all_data.csv file.
+  --nm NM     Number of the latest models we want to save for the test time.
+              Default value = 5.
+```
+The outputs generetaed by model_train.py are:
+
+* Binary conversion of .png ROI files
+* ###_all_features.csv saved in ./train_results directory by default. ### is the date/time when the executable was run. This file contains four columns: TCGA ID (TCGA IDs of each patient), indexes (allocated index for each patient), censored (censoring status of each patient: 0 for NOT censored, 1 for censored), Survival months (survival time of each patient in months).
```
Comment (cooperlab) [inline README.md]: This is just basically duplicating the input, but only for the patients used in training?

File: README.md [hunk 9]
```diff
@@ -28,3 +28,127 @@ After matching the slide filenames against the sample IDs from the clinical data
...
+* ###_training_loss.txt saved in ./train_results directory by default. ### is the date/time when the executable was run. This file contains one column which represents the training loss value at each epoch.
```
Comment (cooperlab) [inline README.md]: What is loss - likelihood or c-index?

File: README.md [hunk 10]
```diff
@@ -28,3 +28,127 @@ After matching the slide filenames against the sample IDs from the clinical data
...
+```
+usage: model_train.py [-h] [-m M] [-p P] [-f F] [-i I] [-r R] [-t T] [-d D]
+                      [--lr LR] [--me ME] [--kp KP] [--bs BS] [--ic IC]
+                      [--ngf NGF] [--gn GN] [--ns NS] [--nm NM]
+
+Arguments for Training the SCNN/GSCNN model
+
+optional arguments:
+  -h, --help  show this help message and exit
+  -m M        image_only or image_genome; image_only for SCNN and image_genome
```
Comment (cooperlab) [inline README.md]: Can we make these arguments 'SCNN' and 'GSCNN' instead of 'image_only' and 'image_genome'? This is more consistent with the paper.

File: README.md [hunk 11]
```diff
@@ -28,3 +28,127 @@ After matching the slide filenames against the sample IDs from the clinical data
...
+              we want to do the Z normalization. This is the column number in
+              the ./inputs/all_dataset.csv file, starting from 0. Default
+              value = 6; Note: The columns we want to normalize should get assigned
+              as the last columns of the all_data.csv file
+  --nm NM     number epochs for model averaging. Default value = 5.
+``` 
+The outputs generetaed by model_test.py are:
+ 
+* Input binary files saved in ./tmp folder by default
+* ###_testset_cindex.txt saved in the folder specified by the -r argument. ### is the date/time when the executable was run. This file contains the final C-Index of the model during testing. As described in the paper, this C-Index represents the accuracy of model-averaged risks (the last 5 models by default that are saved during training). [See the *Methods* section for more details](http://www.pnas.org/content/early/2018/03/09/1717139115).
```
Comment (cooperlab) [inline README.md]: I modified this part significantly - please check for accuracy.

File: README.md [hunk 12]
```diff
@@ -28,3 +28,127 @@ After matching the slide filenames against the sample IDs from the clinical data
...
+* ###_testset_patients_risks.csv saved in ./train_results directory by default. ### is the date/time when the executable was run. This file contains four columns that represent: indexes (allocated index of each testset patient), censored (censoring status for each testset patient: 0 for NOT censored, 1 for censored), survivals (survival time for each testset patient in months), risks (predicted risk values by the model for each testset patient)
```
Comment (cooperlab) [inline README.md]: This is just replicating the input, but keeping only test patients?

Comment (cooperlab) [reply README.md]: 'risks' are the model-averaged risks?

File: README.md (hunk 3 above)
Comment (pooyam) [reply README.md]: Right, I eliminated it from the arguments' list.

File: README.md (hunk 1 above)
Comment (pooyam) [reply README.md]: Fixed it inside the code.

File: README.md (hunk 4 above)
Comment (pooyam) [reply README.md]: Done

File: README.md (hunk 5 above)
Comment (pooyam) [reply README.md]: I changed it to the 'Path to save the trained models and their weights/biases/parameters values'

File: README.md (hunk 6 above)
Comment (pooyam) [reply README.md]: I made it more clear at the "Executables for training and testing models" section.

File: README.md (hunk 7 above)
Comment (pooyam) [reply README.md]: I fixed it inside the code and eliminated the normalization arguments.

File: README.md (hunk 8 above)
Comment (pooyam) [reply README.md]: This is used just for double checking to see if the feature values are correctly converted inside the code. Please let me know if we don't need ###_all_features.csv file as an output.

File: README.md (hunk 9 above)
Comment (pooyam) [reply README.md]: These loss values are the summation of negative partial log likelihood and weight decay at each epoch (explanation inside the paper). I added it to the documentation to make it more clear.

File: README.md (hunk 10 above)
Comment (pooyam) [reply README.md]: Done

File: README.md (hunk 11 above)
Comment (pooyam) [reply README.md]: Checked- I just made a minor change to make it more clear.
